### ISSUE TITLE
How is network byte order handled.

### ISSUE THREAD

**Issue by kbandla** [kbandla/dpkt#60]
_From [rudebo...@gmail.com](https://code.google.com/u/108270286804570581450/) on March 28, 2010 18:18:54_

Hello, im creating an arp broadcast packet like this:

import dpkt
import socket
import binascii

def eth_aton(buffer):
        sp = buffer.split(':')
        buffer = ''.join(sp)
        return binascii.unhexlify(buffer)

arp = dpkt.arp.ARP()
arp.sha=eth_aton('00:24:8c:5b:11:ec')
arp.spa=socket.inet_aton('192.168.1.3')
arp.tha=eth_aton('00:00:00:00:00:00')  
arp.tpa=socket.inet_aton('192.168.1.4')
arp.op=dpkt.arp.ARP_OP_REQUEST  
eth=dpkt.ethernet.Ethernet()
eth.src=arp.sha       
eth.dst=eth_aton('ff:ff:ff:ff:ff:ff')           
eth.data=arp                      
eth.type=dpkt.ethernet.ETH_TYPE_ARP                      

And Im able to send the packet with a PF_PACKET RAW socket:

s = socket.socket(socket.PF_PACKET, socket.SOCK_RAW) 
s.bind((device, dpkt.ethernet.ETH_TYPE_ARP))
s.send(str(eth))            

Everyhting works great and Im able to get a reply, my question is How is
the network byte order handled? dpkt or the BSD socket handles it? I've
tried it in systems with different endiannesses and it works fine.

_Original issue: http://code.google.com/p/dpkt/issues/detail?id=33_

**Comment by saylenty** [comment]
Let's have a look at the "dpkt" source code.
There are a lot of classes that allows us to work with many type of the packets: TCP, UDP, HTTP, Radius and the like.
All of them inherits from dpkt.Packet class which actually determines byte order.
Here is the answer (from dpkt.py):
-----//------
class Packet(object):

    Base packet class, with metaclass magic to generate members from self.hdr.
    hdr should be defined as a list of (name, structfmt, default) tuples
    __byte_order__ can be set to override the default ('>')
-----//------
__byte_order__ field determines either big-endian or little-endian byte order should be used.
The default byte order for dpkt is set in _MetaPacket class line 37:
-----//------

    t.__hdr_fmt__ = getattr(t, 'byte_order', '>') + ''.join([x[1] for x in st])
-----//------
To sum up, __"dpkt" takes care about byte ordering__ and it uses big-endian (which is default network byte order) format as a default byte order.

Some more info could be found here:

    https://en.wikipedia.org/wiki/Endianness
    http://betterexplained.com/articles/understanding-big-and-little-endian-byte-order/
    Official python documentation: https://docs.python.org/2/library/struct.html#struct-alignment

@kbandla or @brifordwylie please, have a look at the answer and close the issue if the answer is full and correct.

**Comment by kbandla** [comment]
Yup this is the right answer. I tagged this issue with `documentation` so that we can add this to the docs later.
